language-hask-source/src/Language/Hask/Parsed/AST.hs

```diff
@@ -0,0 +1,575 @@
+    :: Expression -> NonEmpty Qualifier -> Expression
+  -- | Record construction.
+  --
+  -- > $1 { $_ = $_, ... }
+  RecordConstruction
+    :: ConstructorName -> [(VariableName, Expression)] -> Expression
+  -- | Record update.
+  --
+  -- > $1 { $_ = $_, ... }
+  RecordUpdate
+    :: Expression -> NonEmpty (VariableName, Expression) -> Expression
+
+  -- | Unit constructor.
+  --
+  -- > ()
+  Unit
+    :: Value any
+  -- | List literal.
+  --
+  -- > [$_, ...]
+  List
+    :: [Value any] -> Value any
+  -- | Tuple literal.
+  --
+  -- > ($_, $_, ...)
+  Tuple -- TODO: some better approach?
+    :: (Value any, NonEmpty (Value any)) -> Value any
+  Integer
+    :: Integer -> Value any
+  Float
+    :: Rational -> Value any
+  -- | Character literal.
+  --
+  -- > '$1'
+  Char
+    :: Char -> Value any
+  -- | String literal.
+  --
+  -- > "$1"
+  String
+    :: Text -> Value any
+  -- | Wildcard pattern / hole expression.
+  --
+  -- > _
+  Wildcard
+    :: Value any
+  -- | Parenthesized value.
+  --
+  -- > ($1)
+  Parens
+    :: Value any -> Value any
+
+  -- | "n+k" pattern.
+  --
+  -- > $1 + $2
+  Successor
+    :: VariableName -> Integer -> Pattern
+  -- | Negative literal pattern.
+  --
+  -- > - $1
+  NegativeLiteral
+    :: Either Natural Rational -> Pattern
+  -- | "as" pattern, binding variable name to same value as following pattern.
+  --
+  -- > $1 @ $2
+  As
+    :: VariableName -> Pattern -> Pattern
+  -- | Constructor pattern.
+  --
+  -- > ($_ `$2` $_) $_ ...
+  --
+  -- or
+  --
+  -- > ($_ $2 $_) $_ ...
+  --
+  -- or
+  --
+  -- > $2 $_ ...
+  --
+  -- or
+  --
+  -- > $2 { $_ = $_, ... }
+  ConstructorPattern -- TODO: some better approach?
+    :: Bool
+    -- ^ Is constructor name positioned as infix?
+    -> ConstructorName
+    -> Either [Pattern] [(VariableName, Pattern)]
+    -- ^ Pattern arguments or record syntax.
+    -> Pattern
+  -- | Irrefutable pattern.
+  --
+  -- > ~ $1
+  Irrefutable
+    :: Pattern -> Pattern
+
+deriving stock instance Show (Value any)
+
+-- | Parsed statement in @do@ block.
+data Statement
+  = Then Expression
+    -- ^ Simple expression used as a statement.
+  | Bind Pattern Expression
+    -- ^ Binding of expression result to some pattern.
+    --
+    -- > $1 <- $2
+  | Let [Binding]
+    -- ^ "let" statement binding some values.
+    --
+    -- > let $_ = $_, ...
+  | EmptyStatement
+    -- ^ Empty statement - e.g. redundant @;@.
+  deriving stock Show
+
+-- | Parsed qualifier in list comprehension.
+data Qualifier
+  = Generator Pattern Expression
+    -- ^ Generator, binding pattern to some list expression.
+    --
+    -- > $1 <- $2
+  | QualifierLet [Binding]
+    -- ^ "let" qualifier binding some values.
+    --
+    -- let $_ = $_, ...
+  | Guard Expression
+    -- ^ Boolean expression used as a guard.
+  deriving stock Show
+
+-------------------------------------------------------------------------------
+-- | Parsed type.
+data Type
+  = Constrainted Type Type
```

Comment:
Same as above, could we have something more well-typed here, e.g. I would want something like:

```haskell
data Constraint = Constraint
  { constaintName :: Name
  , tyVars :: [Vars]
  , kind :: Kind
  } -- unsure exactly what should go in this record

Constrainted [(Constraint, [Vars])] Type -- Constraint [(C1, [v1,...,vk])] Ty <-> C1 v1 .. v k => Ty
```
What do you think?

Reply:
Don't you want a kind for each `Vars`?